draft-nygren-httpbis-httpssvc.md

```diff
@@ -0,0 +1,791 @@
+records SHOULD be limited to (8?) prior to reaching terminal address
+records.
+
+The SvcFieldValue in this form SHOULD be an empty string and
+clients MUST ignore its contents.
+
+
+## HTTPSSVC records: alternative service form
+
+When SvcRecordType is 1, the combination of SvcDomainName and
+SvcFieldValue within each resource record associates an Alternative
+Service Field Value with an origin.
+
+The SvcFieldValue of the HTTPSSVC resource record contains an Alt-Svc
+Field Value, exactly as defined in Section 4 of {{!AltSvc}},
+but with the uri-host moved to the SvcDomainName field.
+
+For example, if the operator of https://www.example.com
+intends to include an HTTP response header like
+
+ Alt-Svc: hq="svc.example.net:8003"; ma=3600, \
+          h2="svc.example.net:8002"; ma=3600
+
+They would also publish an HTTPSSVC DNS RRSet like
+
+ www.example.com. 3600 IN HTTPSSVC 1 svc.example.net. "hq=\\":8003\\" pri=0"
+                          HTTPSSVC 1 svc.example.net. "h2=\\":8002\\" pri=1"
+
+This data type can be represented as an Unknown RR as described in
+{{!RFC3597}}:
+
+ www.example.com. 3600 IN TYPE??? \\# TBD:WRITEME
+
+This construction is intended to be extensible in two ways.  First,
+any extensions that are made to the Alt-Svc format for transmission
+over HTTPS are also applicable here, unless expressly mentioned
+otherwise.
+
+Second, by defining a way to map non-HTTPS schemes and non-default
+ports we provide a way for the HTTPSSVC to be used for them as needed.
+However, by using the origin name for the RRName for scheme https and
+port 443 we allow HTTPSSVC records to be included at the end of CNAME
+chains for existing site implementations without requiring changes in
+the zone containing the origin.
+
+
+# Differences from Alt-Svc as transmitted over HTTP
+
+Publishing an alternative services form HTTPSSVC record in DNS is
+intended to be equivalent to transmitting this field value over HTTPS,
+and receiving an HTTPSSVC record is intended to be equivalent to
+receiving this field value over HTTPS.  However, there are some small
+differences in the intended client and server behavior.
+
+## Omitting Max Age and Persist
+
+When publishing an HTTPSSVC record in DNS, server operators MUST omit the
+"ma" parameter, which encodes the "max age" (i.e. expiration time) of
+an Alt-Svc Field Value.  Instead, server operators SHOULD encode the
+expiration time in the DNS TTL, and MUST NOT set a TTL longer than the
+intended "max age".
+
+When receiving an HTTPSSVC record, clients MAY synthesize a new "ma"
+parameter from the DNS TTL, in order to interoperate with Alt-Svc processing
+subsystems.
+
+When publishing an HTTPSSVC record, server operators MUST omit the
+"persist" parameter, which indicates whether the client should use
+this record on other network paths.  When receiving an HTTPSSVC record,
+clients MUST discard any records that contain a "persist" flag.
+Disabling persistence is important to prevent an adversary on one
```

Comment:
I’m dubious that omitting persist actually helps.  Once the client has connected to the server, if the server possesses a valid certificate, then it simply supplies the persistent record as an HTTP header.  If the server doesn’t possess a valid certificate, then the persistent entry will fail to connect anyway.

Reply:
The concern here is about an active attacker on the local network who is also a pervasive passive adversary.  This attacker could inject a fraudulent DNS record that would then permit them to track that user even after they leave the network.  The adversary in this case is not the server operator.

For example, suppose there is a server operator who doesn't use Alt-Svc at all, and who serves the same domain on all IPs of a /64.  An adversary who can inject a long-TTL DNS record can pin a user to one of those IPs, and then reidentify that user across the internet whenever they visit that domain.  This can (and should) be solved by clearing the client's DNS cache when moving between networks.  Since DNS data in this case can be cached in the Alt-Svc cache, we need "persist=0" to get equivalent behavior.

Reply:
Clarified in #15.

Reply:
The DoS issue is solved by Alt-Svc failing back to a direct connection if alternatives fail.  But the tracking argument is valid -- the attacker could just relay the packets to the real server, meaning the TLS handshake will complete (albeit with TLS RTT exceeding TCP's RTT from the client's point of view).

Reply:
Good point, that makes this attack even easier.